analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt

```diff
@@ -211,22 +211,32 @@ class SpdxDocumentFile(
             spdxDocument.packages
         }
 
-        val packages = nonProjectPackages.mapTo(sortedSetOf()) { pkg ->
+        fun PackageReference.getPackageReferenceIds(): List<Identifier> =
```

Comment:
This code is pretty functional, but probably not that efficient. I would propose using a mutable set (a set is preferable over a list because it is later used for contains checks) and do something as the following (not syntax checked):

```
fun PackageReference.getPackageReferenceIds(ids: MutableSet<Identifier>) {
  ids += id
  dependencies.forEach { it.getPackageReferenceIds(ids) }
}

fun Scope.getPackageReferenceIds(ids: MutableSet<Identifier>) {
  dependencies.forEach { id.getPackageReferenceIds(ids) }
}

val packagesInScope = project.scopes.fold(mutableSetOf<Identifier>()) { ids, scope ->
  scope.getPackageReferenceIds(ids)
  ids
}

```
Also not that nice, but this approach prevents the creation of lots of temporary collections.

Reply:
Another approach is to define a set for all packages before as done for [Bundler](https://github.com/oss-review-toolkit/ort/blob/952979d02619c9a5328a28f92e07cec84a9a3173/analyzer/src/main/kotlin/managers/Bundler.kt#L99) and once you parse a dependency, you can add it in this [moment](https://github.com/oss-review-toolkit/ort/blob/952979d02619c9a5328a28f92e07cec84a9a3173/analyzer/src/main/kotlin/managers/Bundler.kt#L165-L176). The advantage is that you avoid the contains check and that you do not have to do the recursion again. In our case that would be the `getDependencies` function.

Reply:
I chose to implement @KorSin 's approach, therefore can not apply @oheger-bosch approach anymore since that code was removed